File: 프로그래머스/도넛과 막대 그래프.py
def solution(edges):
    answer = [0,0,0,0]
    nodeCnt={}
    
    for e in edges:
        a=e[0]
        b=e[1]
        if not nodeCnt.get(a):
            nodeCnt[a]=[0,0]
        if not nodeCnt.get(b):
            nodeCnt[b]=[0,0]
        nodeCnt[a][0]+=1
        nodeCnt[b][1]+=1
        
    for k,v in nodeCnt.items():
        if v[0]>=2 and v[1]==0:
            answer[0]=k
        elif v[0]>=2 and v[1]>=2:
            answer[3]+=1
        elif v[0]==0 and v[1]>=1:
            answer[2]+=1
    answer[1]=nodeCnt[answer[0]][0]-answer[2]-answer[3]
            
    return answer

File: 프로그래머스/도넛과 막대 그래프.java
import java.util.*;
class Solution {
    public int[] solution(int[][] edges) {
        int[] answer = {0,0,0,0};
        Map<Integer,int[]> map=new HashMap<>();
        
        for(int[] e:edges){
            int a=e[0];
            int b=e[1];
            
            if (!map.containsKey(a)){
                map.put(a,new int[]{0,0});
            }
            if (!map.containsKey(b)){
                map.put(b,new int[]{0,0});
            }
            map.get(a)[0]++;
            map.get(b)[1]++;
        }
        
        map.forEach((k,v)->{
            if (v[0]>=2 && v[1]==0){
                answer[0]=k;
            }else if(v[0]==0 && v[1]>=1){
                answer[2]++;
            }else if(v[0]>=2 && v[1]>=2){
                answer[3]++;
            }
        });
        answer[1]=map.get(answer[0])[0]-answer[2]-answer[3];
        return answer;
    }
}